Prevent malformed prefixes caused by missing parameters

jayyslee opened the pull request:
When prefix parameters are expected, but do not exist, the resulting prefix may contain repeating slashes and end with a slash. If the prefix ends with a slash, the hostname is dropped from the request URL.

To fix this, I collapsed repeating slashes and removed any trailing slashes.

pickle27 commented:
thanks for sending a patch, I see you've added a test but could you also add a quick description of the issue this is fixing to make things easier on the maintainers?

Thanks!

jayyslee commented:
I've left a description. Let me know if you need any additional information.

pickle27 commented:
thanks! I've noticed this issue before myself. Let me give a review, maybe @dylanahsmith can take a look as well

pickle27 commented on test/activeresource_test.py:
```diff
@@ -234,6 +234,14 @@ def test_set_prefix_source(self):
+        self.person.prefix_source = '/${store_type}/${store_id}/'
+        nobody = self.person.find(name='Ralph')
+        self.assertEqual([], nobody)
+
```
can we have a test with multiple `/` ?

dylanahsmith commented on pyactiveresource/activeresource.py:
```diff
@@ -738,7 +738,12 @@ def _prefix(cls, options=None):
+
+        prefix = re.sub(r'/+', '/', prefix)
+        if prefix.endswith('/'):
+            prefix = prefix[:-1]
```
I think we can remove the `path = re.sub('/$', '', cls.prefix_source)` line since this does the same thing.

dylanahsmith commented:
> When prefix parameters are expected, but do not exist, the resulting prefix may contain repeating slashes and end with a slash.

I thought the point of the prefix parameters was to have required parts of the URL, which it is usually a bug to be missing.  Also, I am surprised that the server doesn't normalize the url to ignore redundant slashes.

> If the prefix ends with a slash, the hostname is dropped from the request URL.

I don't understand what you mean.  Why would the hostname be dropped from the URL in this case?

dylanahsmith commented:
This isn't something that is done in [ActiveResource](https://github.com/rails/activeresource), so I don't see why we need to do it in pyactiveresource, since it is just a port of active resource.  Perhaps you could propose the change there first.

jayyslee commented:
Error: Unknown request GET http://people.json?name=Ralph

That's what happens when no prefix parameters exist, and the prefix ends up being '/'. I've updated the test to reflect this. Without the fix, the test will fail with that error.

In my opinion, the line that I removed, 737, suggests that the prefix should not be ending with a slash. With missing parameters, it is quite possible to have a prefix like '///', therefore I collapse them and remove the last slash.

dylanahsmith commented:
The problem isn't that the prefix ends with a slash, it looks like it is that when the prefix starts with two slashes that it interprets the following path component as the host, which overrides the host in the site when doing `urllib.parse.urljoin(self.site, path)` in Connection._open.

Given that, I think all that is needed is the following change:

```diff
diff --git a/pyactiveresource/activeresource.py b/pyactiveresource/activeresource.py
index d6bac84..4121d16 100644
--- a/pyactiveresource/activeresource.py
+++ b/pyactiveresource/activeresource.py
@@ -738,7 +738,8 @@ class ActiveResource(six.with_metaclass(ResourceMeta, object)):
         template = Template(path)
         keys = cls._prefix_parameters()
         options = dict([(k, options.get(k, '')) for k in keys])
-        return template.safe_substitute(options)
+        prefix = template.safe_substitute(options)
+        return re.sub('^/+', '', prefix)

     # Public instance methods
     def to_dict(self):
```

along with your regression test.

jayyslee commented:
Sounds good to me

dylanahsmith commented:
Thanks for helping us fix this bug!
